velox/exec/tests/HashJoinWithCacheTest.cpp

```diff
@@ -0,0 +1,543 @@
+        .queryCtx(queryCtx)
+        .maxDrivers(numDrivers)
+        .assertResults(
+            "SELECT t.t_k, t.t_v, u.u_k, u.u_v FROM t, u WHERE t.t_k = u.u_k");
+  };
+
+  // Run 10 threads concurrently, each executing 5 tasks sequentially.
+  constexpr int kNumThreads = 10;
+  constexpr int kTasksPerThread = 5;
+  constexpr int kTotalTasks = kNumThreads * kTasksPerThread;
+
+  // Each thread maintains its own local vector of tasks.
+  std::vector<std::vector<std::shared_ptr<Task>>> threadTasks(kNumThreads);
+  std::vector<std::thread> threads;
+  threads.reserve(kNumThreads);
+
+  // Launch all threads at once.
+  for (int i = 0; i < kNumThreads; ++i) {
+    threads.emplace_back([&, i]() {
+      // Each thread runs multiple tasks sequentially into its local vector.
+      threadTasks[i].reserve(kTasksPerThread);
+      for (int j = 0; j < kTasksPerThread; ++j) {
+        threadTasks[i].push_back(runTask());
+      }
+    });
+  }
+
+  // Wait for all threads to complete.
+  for (auto& thread : threads) {
+    thread.join();
+  }
+
+  // Merge all thread-local task vectors into a single vector.
+  std::vector<std::shared_ptr<Task>> allTasks;
+  allTasks.reserve(kTotalTasks);
+  for (auto& tasks : threadTasks) {
+    for (auto& task : tasks) {
+      allTasks.push_back(std::move(task));
+    }
+  }
+
+  ASSERT_EQ(allTasks.size(), kTotalTasks);
+
+  // Collect stats from all tasks.
+  int totalCacheMisses = 0;
+  int totalCacheHits = 0;
+
+  for (const auto& task : allTasks) {
+    auto opStats = toOperatorStats(task->taskStats());
+    ASSERT_EQ(opStats.count("HashBuild"), 1);
+    auto& hashBuildStats = opStats.at("HashBuild");
+
+    if (hashBuildStats.runtimeStats.count(BaseHashTable::kHashTableCacheMiss)) {
+      totalCacheMisses +=
+          hashBuildStats.runtimeStats.at(BaseHashTable::kHashTableCacheMiss)
+              .count;
+    }
+    if (hashBuildStats.runtimeStats.count(BaseHashTable::kHashTableCacheHit)) {
+      totalCacheHits +=
+          hashBuildStats.runtimeStats.at(BaseHashTable::kHashTableCacheHit)
+              .count;
+    }
+  }
+
+  // Exactly one task should build (cache miss) and all others should reuse
+  // (cache hits).
+  EXPECT_EQ(totalCacheMisses, 1)
+      << "Exactly one task should report a cache miss (the builder)";
+  EXPECT_EQ(totalCacheHits, kTotalTasks - 1)
+      << "All other tasks should report cache hits";
+
+  // Clean up cache entry before tasks are destroyed.
+  const auto cacheKey = fmt::format("{}:{}", queryId, joinNodeId);
+  HashTableCache::instance()->drop(cacheKey);
+}
+
+// Tests that HashProbe cannot spill when using a cached hash table.
+// When useHashTableCache() is true, HashProbe::canSpill() returns false
+// because spilling would clear the cached table and corrupt it for other tasks.
+// This test verifies that HashProbe completes without spilling.
+TEST_F(HashJoinWithCacheTest, probeCannotSpillWithCachedTable) {
+  const std::string queryId =
```

Comment:
This test should use testvalue to check if canSpill() is set to false for both build and probe. Rename the test case to joinWithCachedTableCanNotSpill? This test will simply succeed no matter spill enabled or not?